### PR TITLE
Add state machine features

### PR DESCRIPTION
**Pull request by j-mueller**
* Adds 'Language.PlutusTx.StateMachine', a module that implements state machines in terms of Plutus transactions, incl. checks of validator and redeemer scripts
* Adds an extended guessing game contract that implements tokens for authorisation and contract state
* Extends the standard library with some useful things

### CODE REVIEW

**Comment by j-mueller** [comment]
By the way, the tests for the state machine take quite a long time to run:

```
state machine tests
    run a successful game trace:               Script size: 51043
OK (12.47s)
    run a 2nd successful game trace:           Script size: 51043
OK (6.95s)
    run a failed trace:                        Script size: 51043
OK (6.14s)
```

I'm not sure why that is!

**Comment by michaelpj** [comment]
@m-alvarez for state machine test slowness

**Comment by j-mueller** [comment]
> @m-alvarez for state machine test slowness

That's not using the state machine, they're unit tests with a single trace

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    let 
+
+        -- | Check whether a 'ClearString' is the preimage of a
+        --   'HashedString'
+        checkGuess :: HashedString -> ClearString -> Bool
+        checkGuess (HashedString actual) (ClearString gss) =
+            $$(P.equalsByteString) actual ($$(P.sha2_256) gss)
+            
+        -- | Given a 'TokeName', get the value that contains 
+        --   exactly one token of that name
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: ... in this script's currency.

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      ForgeToken TokenName
+    -- ^ Forge the "guess" token
+    | Guess ClearString HashedString
+    -- ^ Make a guess and lock the remaining funds using a new secret word.
+
+PlutusTx.makeLift ''GameInput
+
+gameValidator :: ValidatorScript
+gameValidator = ValidatorScript ($$(Ledger.compileScript [||
+    \(ds :: (GameState, Maybe GameInput)) (vs :: (GameState, Maybe GameInput)) (p :: PendingTx) ->
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs]: Maybe we should also define a type alias for "state machine validators" in `StateMachine.hs`?

**Comment by j-mueller** [comment]
The contract for the demo will be in a gist, so we can make cosmetic changes later on. I just wanted to get the library functions and the unit tests in.

**Pull request merged**